devito/mpi/distributed.py

```diff
@@ -600,17 +616,53 @@ class CustomTopology(tuple):
     """
 
     def __new__(cls, items, input_comm):
+        # Keep track of nstars and already defined decompositions
         nstars = len([i for i in items if i == '*'])
-        if nstars > 0:
-            if input_comm.size % nstars != 0:
-                raise ValueError("Invalid `topology` for given nprocs")
-            if any(i not in ('*', 1) for i in items):
-                raise ValueError("Custom topology must be only 1 or *")
-
-            v = input_comm.size // nstars
-            processed = [i if i == 1 else v for i in items]
-        else:
+
+        # If no stars exist we are ready
+        if nstars == 0:
             processed = items
+        else:
+            # Init decomposition list
+            processed = [1] * len(items)
+
+            # Get star and integer indices
+            int_pos = [i for i, item in enumerate(items) if isinstance(item, int)]
+            int_vals = [item for item in items if isinstance(item, int)]
+            star_pos = [i for i, item in enumerate(items) if not isinstance(item, int)]
+
+            # Decompose the processes remaining for allocation to prime factors
+            alloc_procs = np.prod([i for i in items if i != '*'])
+            remprocs = int(input_comm.size // alloc_procs)
+            prime_factors = primefactors(remprocs)
+
+            star_i = -1
+            dd_list = [1] * nstars
+
+            # Start by using the max prime factor at the first starred position,
+            # then cyclically-iteratively decompose as evenly as possible until
+            # decomposing to the number of `remprocs`
+            while remprocs != 1:
+                star_i = star_i + 1
+                star_i = star_i % nstars
+                prime_factors = primefactors(remprocs)
+                dd_list[star_i] = dd_list[star_i]*max(prime_factors)
+                remprocs = remprocs // max(prime_factors)
+
+            if int_pos:
+                for index, value in zip(int_pos, int_vals):
+                    processed[index] = value
+
+            if dd_list:
+                for index, value in zip(star_pos, dd_list):
+                    processed[index] = value
+
+        # Final check that topology matches the communicator size
+        try:
+            assert np.prod(processed) == input_comm.size
+        except:
+            raise ValueError("Invalid `topology`", processed, " for given nprocs:",
```

Comment:
if it's an assert, why bothering with this? Just leave the assert alone and that's it

Reply:
I think it will be more helpful as a message to the user?
I can drop if you feel so